add correspondence grouping algorithm into recognition node

lianjunli opened the pull request:
PCL tutorial for correspondence grouping can be found[ here](http://pointclouds.org/documentation/tutorials/correspondence_grouping.php#correspondence-grouping).

lianjunli committed:
add Correspondence Grouping algorithm for recognition node

lianjunli committed:
add Correspondence Grouping algorithm for recognition node

lianjunli committed:
add Correspondence Grouping algorithm for recognition node

jrgnicho commented on gilbreth_perception/src/recognition_node.cpp:
```diff
@@ -119,83 +124,185 @@ void loadModel(std::vector<std::vector<double> > &pick_pose, std::vector<std::st
   }
 }
 
 void cloudCallBack(const sensor_msgs::PointCloud2ConstPtr &cloud_msg, const tf::TransformListener &listener,
```
Rather than passing this many arguments in one function would it be best to make this function into a class method and set the arguments as class members?

lianjunli replied:
Yes, I was thinking about that last Friday, I am going to change it.

jrgnicho commented on gilbreth_perception/src/recognition_node.cpp:
```diff
@@ -67,10 +70,12 @@ void loadParameter() {
-               std::vector<pcl::PointCloud<PointType>::Ptr> &model_list, std::vector<pcl::PointCloud<pcl::FPFHSignature33>::Ptr> &model_features_list) {
+void loadModel(std::vector<std::vector<double>> &pick_pose, std::vector<std::string> &model_name,
+               std::vector<pcl::PointCloud<PointType>::Ptr> &model_list, std::vector<pcl::PointCloud<pcl::FPFHSignature33>::Ptr> &model_features_list,
+               std::vector<pcl::PointCloud<pcl::SHOT352>::Ptr> &model_descriptor_list, std::vector<pcl::PointCloud<pcl::ReferenceFrame>::Ptr> &model_rf_list) {
```
To make the code cleaner, you create a ```loadICPConfig``` and ```loadCGConfig``` and call one or the other depending on what's chosen.

jrgnicho commented on gilbreth_perception/src/recognition_node.cpp:
```diff
@@ -42,14 +43,16 @@ float min_sample_distance(0.025f);
 bool visualizer(false);
+bool icp(true);
+float cg_size_(0.05f);
+float cg_thresh_(8.0f);
```
Are these new variables configurable from the yaml?

jrgnicho reviewed (changes requested):
Just a few modifications, thanks

lianjunli committed:
Wrap all functions and variables into a class;
Using tf information in Michael's environment;
Model number increased to 5;
Tuning parameters to fit the new environment;
Delete pcl viewer;

jrgnicho commented on gilbreth_perception/config/parameters.yaml:
```diff
@@ -1,13 +1,17 @@
-  z: [0.0, 1.19]
+  x: [-0.22543, 0.192618]
+  y: [-0.2874, 0.283763]
+  z: [0.0, 0.7162]
```
Are these parameters applied relative to the camera coordinate frame?

jrgnicho commented on gilbreth_perception/src/segmentation_node.cpp:
```diff
@@ -55,45 +58,50 @@ void cloudCb(const sensor_msgs::PointCloud2ConstPtr &cloud_msg) {
+          std::cerr << "Publish cluster " << i + 1 << std::endl;
+        }
+      }
+    }
```
In general, I try to avoid too many levels of nested if statements,  whenever possible I prefer to do as follows:
```
if (condition1_valid == true)
{
  // ... do something meaningful
}
else
{
  // ... error handling procedure if applicable
  return;
}

if (condition2_valid == true)
{
  // ... do something meaningful
}
else
{
  // ... error handling procedure if applicable
  return;
}
```
In my opinion it helps with code readability, however no change is needed on this one.
